diff --git a/src/tsutil/CMakeLists.txt b/src/tsutil/CMakeLists.txt
index 0d17737d4..c604af08c 100644
--- a/src/tsutil/CMakeLists.txt
+++ b/src/tsutil/CMakeLists.txt
@@ -16,6 +16,7 @@
 #######################
 
 set(TSUTIL_PUBLIC_HEADERS
+    ${PROJECT_SOURCE_DIR}/include/tsutil/Assert.h
     ${PROJECT_SOURCE_DIR}/include/tsutil/Metrics.h
     ${PROJECT_SOURCE_DIR}/include/tsutil/SourceLocation.h
     ${PROJECT_SOURCE_DIR}/include/tsutil/DbgCtl.h
